File: go/tracked.go
package headgate

// Structured background work owned by one handler attempt. A bare goroutine is outside
// the runner's lifecycle; Track registers it so success and graceful shutdown wait for
// it, while lease loss and forced shutdown cancel its context.

import (
	"context"
	"errors"
	"fmt"
	"sync"
)

var (
	ErrTaskTrackerUnavailable = errors.New("headgate: tracked tasks are only available inside a handler")
	ErrTaskTrackerClosed      = errors.New("headgate: job attempt is no longer accepting tracked tasks")
)

type taskTrackerContextKey struct{}

type taskTracker struct {
	mu        sync.Mutex
	accepting bool
	ctx       context.Context
	cancel    context.CancelFunc
	wg        sync.WaitGroup
	firstErr  error
}

func withTaskTracker(parent context.Context) (context.Context, *taskTracker) {
	base, cancel := context.WithCancel(parent)
	tracker := &taskTracker{accepting: true, cancel: cancel}
	bound := context.WithValue(base, taskTrackerContextKey{}, tracker)
	tracker.ctx = bound
	return bound, tracker
}

// Track starts work concurrently and attaches it to the current job attempt. The work
// receives the handler's cancellation/deadline context and may itself call Track. The
// first tracked error fails the attempt after all sibling work has been cancelled and
// joined. Calling Track outside dispatch, or after the handler has returned, is an
// explicit error—there is no process-global tracker.
func Track(ctx context.Context, work func(context.Context) error) error {
	if work == nil {
		return errors.New("headgate: tracked task function is nil")
	}
	tracker, ok := ctx.Value(taskTrackerContextKey{}).(*taskTracker)
	if !ok || tracker == nil {
		return ErrTaskTrackerUnavailable
	}
	return tracker.spawn(work)
}

func (tracker *taskTracker) spawn(work func(context.Context) error) error {
	tracker.mu.Lock()
	defer tracker.mu.Unlock()
	if !tracker.accepting {
		return ErrTaskTrackerClosed
	}
	// Add happens under the same lock that closes admission before Wait, preventing
	// the WaitGroup Add/Wait zero-counter race.
	tracker.wg.Add(1)
	go func() {
		defer tracker.wg.Done()
		var err error
		defer func() {
			if recovered := recover(); recovered != nil {
				err = fmt.Errorf("panic in tracked task: %v", recovered)
			}
			if err != nil {
				tracker.recordError(err)
			}
		}()
		err = work(tracker.ctx)
	}()
	return nil
}

func (tracker *taskTracker) recordError(err error) {
	tracker.mu.Lock()
	if tracker.firstErr == nil {
		tracker.firstErr = err
		// One child failed, so the attempt cannot complete successfully. Cancel its
		// siblings immediately; wait() below still joins every goroutine.
		tracker.cancel()
	}
	tracker.mu.Unlock()
}

func (tracker *taskTracker) wait() error {
	tracker.close(false)
	tracker.wg.Wait()
	tracker.cancel() // release the derived context after normal completion
	tracker.mu.Lock()
	err := tracker.firstErr
	tracker.mu.Unlock()
	return err
}

func (tracker *taskTracker) cancelAndWait() {
	tracker.close(true)
	tracker.wg.Wait()
}

func (tracker *taskTracker) close(cancel bool) {
	tracker.mu.Lock()
	tracker.accepting = false
	if cancel {
		tracker.cancel()
	}
	tracker.mu.Unlock()
}

File: crates/headgate/src/tracked.rs
//! Structured background work owned by one job attempt.
//!
//! A handler may need to start concurrent work and return before that work has joined.
//! A bare `tokio::spawn` detaches it from the job: the worker can acknowledge or shut
//! down while the future is still producing side effects. `TaskTracker` keeps those
//! futures inside the attempt's lifecycle instead.

use std::any::Any;
use std::future::Future;
use std::sync::{Arc, Mutex};

use headgate_core::BoxError;
use tokio::task::JoinSet;

/// The handler has already returned, so this attempt no longer accepts new tracked
/// work. This commonly means a detached task retained a `JobCtx` and tried to register
/// more work after its owner had begun finishing.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct TrackedTaskClosed;

impl std::fmt::Display for TrackedTaskClosed {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "job attempt is no longer accepting tracked tasks")
    }
}

impl std::error::Error for TrackedTaskClosed {}

struct State {
    accepting: bool,
    tasks: Option<JoinSet<Result<(), BoxError>>>,
}

/// Shared by all clones of one `JobCtx`. The `JoinSet` is deliberately owned rather
/// than a collection of bare `JoinHandle`s: dropping a JoinSet aborts its children,
/// while dropping a JoinHandle detaches them.
pub(crate) struct TaskTracker {
    state: Mutex<State>,
    liveness: Arc<crate::stuck_handler::AttemptLiveness>,
}

pub(crate) enum TrackedFailure {
    Error(BoxError),
    Panic(Box<dyn Any + Send + 'static>),
}

impl TaskTracker {
    pub(crate) fn new(liveness: Arc<crate::stuck_handler::AttemptLiveness>) -> Arc<Self> {
        Arc::new(Self {
            state: Mutex::new(State {
                accepting: true,
                tasks: Some(JoinSet::new()),
            }),
            liveness,
        })
    }

    pub(crate) fn spawn<F>(&self, future: F) -> Result<(), TrackedTaskClosed>
    where
        F: Future<Output = Result<(), BoxError>> + Send + 'static,
    {
        let mut state = self.state.lock().unwrap();
        if !state.accepting {
            return Err(TrackedTaskClosed);
        }
        let Some(tasks) = state.tasks.as_mut() else {
            return Err(TrackedTaskClosed);
        };
        let active = self.liveness.activate();
        tasks.spawn(async move {
            let _active = active;
            future.await
        });
        Ok(())
    }

    /// Stop accepting children and wait for every registered future. The first child
    /// error/panic fails the attempt and aborts its siblings; acknowledging success
    /// while one sibling is known to have failed would be a false completion.
    pub(crate) async fn finish(&self) -> Result<(), TrackedFailure> {
        let Some(mut tasks) = self.take() else {
            return Ok(());
        };
        let mut first = None;
        let mut aborting = false;
        while let Some(joined) = tasks.join_next().await {
            let failure = match joined {
                Ok(Ok(())) => None,
                Ok(Err(error)) => Some(TrackedFailure::Error(error)),
                Err(error) if error.is_panic() => Some(TrackedFailure::Panic(error.into_panic())),
                Err(error) if error.is_cancelled() && aborting => None,
                Err(error) => Some(TrackedFailure::Error(
                    format!("tracked task was cancelled unexpectedly: {error}").into(),
                )),
            };
            if first.is_none()
                && let Some(failure) = failure
            {
                first = Some(failure);
                aborting = true;
                tasks.abort_all();
            }
        }
        match first {
            Some(failure) => Err(failure),
            None => Ok(()),
        }
    }

    /// Synchronous half of cancellation, used before the worker aborts its outer
    /// attempt task. This is load-bearing for a tracked future that owns a `JobCtx`:
    /// without `abort_all`, `JobCtx -> tracker -> future -> JobCtx` would keep detached
    /// work alive after lease loss.
    pub(crate) fn cancel(&self) {
        let mut state = self.state.lock().unwrap();
        state.accepting = false;
        if let Some(tasks) = state.tasks.as_mut() {
            tasks.abort_all();
        }
    }

    /// Cancel and join every child before an ordinary handler error is acknowledged.
    pub(crate) async fn cancel_and_wait(&self) {
        let Some(mut tasks) = self.take() else {
            return;
        };
        tasks.abort_all();
        while tasks.join_next().await.is_some() {}
    }

    fn take(&self) -> Option<JoinSet<Result<(), BoxError>>> {
        let mut state = self.state.lock().unwrap();
        state.accepting = false;
        state.tasks.take()
    }
}

impl Drop for TaskTracker {
    fn drop(&mut self) {
        if let Ok(state) = self.state.get_mut()
            && let Some(tasks) = state.tasks.as_mut()
        {
            tasks.abort_all();
        }
    }
}
